### PR TITLE
QUICKFIX: gracefully handle unexpected unlinks

### PR DESCRIPTION
**Pull request by thousand**

### CODE REVIEW

**Comment by rwjblue** [comment]
This seems good to me (if we get an `ENOENT` that means the file doesn't exist, so we should be "fine"), mind adding a test?

Also, can you describe how we get into the invalid state (just in case there are other issues that need to be addressed)?

File: lib/index.js [hunk 1]
```diff
@@ -13,7 +13,15 @@ var validateSortedUnique = util.validateSortedUnique;
 var ARBITRARY_START_OF_TIME = 0;
 var DEFAULT_DELEGATE = {
   unlink: function(inputPath, outputPath, relativePath) {
-    fs.unlinkSync(outputPath);
+    try {
+      fs.unlinkSync(outputPath);
+    } catch (e) {
+      // QUICKFIX: gracefully handle unexpected unlinks (from broccoli-dependency-funnel)
+      if (typeof e === 'object' && e !== null && e.code = 'ENOENT') {
```
Comment (rwjblue) [inline lib/index.js]: This seems slightly incorrect, I believe you mean to not assign the `e.code` property?

```js
if (typeof e === 'object' && e !== null && e.code === 'ENOENT') { return; }
```

File: lib/index.js [hunk 2]
```diff
@@ -13,7 +13,15 @@ var validateSortedUnique = util.validateSortedUnique;
 var ARBITRARY_START_OF_TIME = 0;
 var DEFAULT_DELEGATE = {
   unlink: function(inputPath, outputPath, relativePath) {
-    fs.unlinkSync(outputPath);
+    try {
+      fs.unlinkSync(outputPath);
+    } catch (e) {
+      // QUICKFIX: gracefully handle unexpected unlinks (from broccoli-dependency-funnel)
```
Comment (rwjblue) [inline lib/index.js]: Probably don't need this comment?

File: lib/index.js (hunk 1 above)
Comment (rwjblue) [reply lib/index.js]: This is what is failing the CI build.

File: lib/index.js (hunk 2 above)
Comment (stefanpenner) [reply lib/index.js]: i wrote this :P

**Comment by thousand** [comment]
Fixed the logic and added a test! Scenario that causes the issue is when the patch indicates an unlink, and the target directory doesn't contain that file. Test indicates the scenario we found where another process is deleting the file between the time the patch is generated and applied.

**Pull request merged**